File: nim/secret-handshake/secret_handshake.nim
import algorithm, bitops

const lookup = block:
  func toCommands(handshake: int): seq[string] =
    if testBit(handshake, 0): result.add("wink")
    if testBit(handshake, 1): result.add("double blink")
    if testBit(handshake, 2): result.add("close your eyes")
    if testBit(handshake, 3): result.add("jump")
    if testBit(handshake, 4): result.reverse

  func genLookup(): array[32, seq[string]] =
    for i in result.low..result.high:
      result[i] = i.toCommands()

  genLookup()

func commands*(handshake: int): seq[string] =
  lookup[handshake mod lookup.len]

File: r/secret-handshake/secret-handshake.R
actions <- c(
  \(commands) { append(commands, "wink") },
  \(commands) { append(commands, "double blink") },
  \(commands) { append(commands, "close your eyes") },
  \(commands) { append(commands, "jump") },
  \(commands) { rev(commands)}
)

handshake <- function(n) {
  commands <- c()
  
  for (index in seq_along(actions)) {
    if (bitwAnd(bitwShiftL(1, index - 1), n) != 0) {
      commands <- actions[[index]](commands)
    }
  }
  
  commands
}
